File: apps/api/src/modules/plugins/plugin.loader.ts
import { readFile, readdir, stat } from 'node:fs/promises';
import { basename, resolve } from 'node:path';
import { pathToFileURL } from 'node:url';
import { Injectable } from 'injectkit';
import { satisfies, validRange } from 'semver';
import { PLUGIN_API_VERSION, pluginManifestSchema, type DeadairPlugin, type PluginManifest } from '@deadair/plugin-sdk';
import type { PluginOrigin, PluginRecord } from './types/plugin.record.js';
import { errorText } from '#modules/shared/error.text.js';

/**
 * Where the loader looks for plugins: the directories shipped with the host
 * (`bundledDirs`) and the operator-installed mount point (`pluginsDir`).
 *
 * Both arrive by constructor injection so the module can wire the
 * `PLUGINS_DIR` env and the bundled list without the loader ever touching
 * `AppConfig` or a module-level constant. `bundledDirs` defaults to empty so a
 * caller that constructs a loader directly (a test, a one-off scan) sees
 * exactly the directory it asked for and nothing else.
 */
@Injectable()
export class PluginLoaderOptions {
    constructor(
        readonly pluginsDir: string,
        readonly bundledDirs: string[] = [],
    ) {}
}

/**
 * Every entry path this process has handed to `import()`, resolved.
 *
 * Module state rather than an instance field because what it records is: Node caches an ES module
 * by URL for the life of the PROCESS, whichever loader asked for it, so a second loader constructed
 * for one scan (the installer validating a staged plugin, say) has to see the first one's imports.
 * Added to BEFORE the import rather than after it, because a module that throws while it evaluates
 * is cached as failed and is just as stale as one that loaded.
 */
const importedEntries = new Set<string>();

/**
 * Whether the code at `entryPath` has already been imported into this process, and so would not
 * be read again if the file changed.
 *
 * This is what lets the installer say when a plugin it just placed needs a restart: new code at a
 * NEW path is read on the next discovery, and new code at a path already imported is not.
 */
export function hasImportedPluginEntry(entryPath: string): boolean {
    return importedEntries.has(resolve(entryPath));
}

/** The `deadair` block a plugin package.json must carry to be a candidate. */
interface PluginPackageJson {
    deadair?: {
        plugin?: unknown;
    };
}

/** Whether `path` resolves to an existing regular file. */
async function fileExists(path: string): Promise<boolean> {
    try {
        return (await stat(path)).isFile();
    } catch {
        return false;
    }
}

/**
 * Discovers plugin candidates on disk and turns each one into a
 * {@link PluginRecord}.
 *
 * Discovery is untrusted-input handling: a plugin directory is arbitrary
 * third-party code that may not parse, may throw at import time, or may lie
 * about its manifest. Every one of those outcomes becomes a quarantined
 * record with error text, never an exception, so a single bad plugin can
 * never take the server's boot down with it.
 *
 * This is the only place in the host that imports plugin code.
 */
@Injectable()
export class PluginLoader {
    constructor(private readonly options: PluginLoaderOptions) {}

    /**
     * Scans the bundled list plus the mounted plugins directory and returns one
     * record per candidate, in that order. Directories that are not plugins at
     * all are skipped silently; broken plugins come back as `failed` records.
     *
     * Never rejects.
     */
    async discover(): Promise<PluginRecord[]> {
        const candidates: { dir: string; origin: PluginOrigin }[] = [
            ...this.options.bundledDirs.map(dir => ({ dir: resolve(dir), origin: 'bundled' as const })),
            ...(await this.scanPluginsDir()).map(dir => ({ dir, origin: 'installed' as const })),
        ];

        const records: PluginRecord[] = [];
        const seenDirs = new Set<string>();
        const seenIds = new Set<string>();

        for (const { dir, origin } of candidates) {
            if (seenDirs.has(dir)) continue;
            seenDirs.add(dir);

            const record = await this.loadCandidate(dir, origin, seenIds);
            if (record) records.push(record);
        }

        return records;
    }

    /**
     * Subdirectories of `pluginsDir`, absolute and sorted for a stable load
     * order. A missing or unreadable plugins directory is normal (nothing has
     * been installed yet), so it yields an empty list rather than an error.
     */
    private async scanPluginsDir(): Promise<string[]> {
        const root = resolve(this.options.pluginsDir);

        let entries;
        try {
            entries = await readdir(root, { withFileTypes: true });
        } catch {
            return [];
        }

        const dirs: string[] = [];
        for (const entry of entries) {
            // The station's own: `PluginPeerLinker` keeps the host's SDK and zod here so an installed
            // plugin can resolve them. It has no package.json and would be skipped anyway, but it is
            // never a plugin, so it is not even looked at.
            if (entry.name === 'node_modules') continue;
            // Never a plugin: the installer stages an upload under `.staging` while it checks it, and a
            // rescan landing in the middle of that must not see half a plugin. The same rule keeps
            // `.DS_Store`, `.git` and an editor's droppings out, which no plugin folder is named like.
            if (entry.name.startsWith('.')) continue;
            const full = resolve(root, entry.name);
            if (entry.isDirectory()) {
                dirs.push(full);
                continue;
            }
            // Symlinked plugin directories are the natural way to develop a
            // plugin against a running server, so follow them.
            if (entry.isSymbolicLink()) {
                try {
                    if ((await stat(full)).isDirectory()) dirs.push(full);
                } catch {
                    // Dangling symlink: not a plugin, nothing to report.
                }
            }
        }

        return dirs.sort();
    }

    /**
     * Turns one directory into a record, or `undefined` when the directory is
     * not a plugin at all.
     *
     * `seenIds` is mutated: the first plugin to claim an id keeps it, and every
     * later claimant is quarantined.
     */
    private async loadCandidate(dir: string, origin: PluginOrigin, seenIds: Set<string>): Promise<PluginRecord | undefined> {
        let entry: string;
        let entryPath: string;
        try {
            const declared = await this.readEntryPath(dir);
            if (declared === undefined) return undefined;
            entry = declared;
            entryPath = resolve(dir, declared);
        } catch (error) {
            return this.quarantine(dir, origin, `unreadable package.json: ${errorText(error)}`);
        }

        // A plugin that declares a built entry it never shipped is the single
        // most common operator mistake (an unbuilt checkout). Naming the entry
        // and the fix beats surfacing a raw ERR_MODULE_NOT_FOUND stack.
        if (!(await fileExists(entryPath))) {
            return this.quarantine(
                dir,
                origin,
                `entry "${entry}" does not exist; the plugin has not been built (run its build, and install dist/ with its package.json)`,
            );
        }

        let module: unknown;
        try {
            importedEntries.add(entryPath);
            module = await import(pathToFileURL(entryPath).href);
        } catch (error) {
            return this.quarantine(dir, origin, `failed to import entry "${entryPath}": ${errorText(error)}`);
        }

        const exported = (module as { default?: unknown }).default;
        if (typeof exported !== 'object' || exported === null) {
            return this.quarantine(dir, origin, 'entry has no default export; a plugin must default-export definePlugin(manifest, factory)');
        }

        const { manifest, factory } = exported as Partial<DeadairPlugin>;
        if (typeof factory !== 'function') {
            return this.quarantine(dir, origin, 'default export has no factory function');
        }

        const parsed = pluginManifestSchema.safeParse(manifest);
        if (!parsed.success) {
            const issues = parsed.error.issues.map(issue => `${issue.path.join('.') || '(root)'}: ${issue.message}`).join('; ');
            return this.quarantine(dir, origin, `invalid manifest: ${issues}`);
        }
        const validated = parsed.data as PluginManifest;

        if (!validRange(validated.apiVersion)) {
            return this.quarantine(dir, origin, `manifest apiVersion "${validated.apiVersion}" is not a valid semver range`, validated.id);
        }
        if (!satisfies(PLUGIN_API_VERSION, validated.apiVersion)) {
            return this.quarantine(
                dir,
                origin,
                `plugin requires plugin API "${validated.apiVersion}" but this host implements ${PLUGIN_API_VERSION}`,
                validated.id,
            );
        }

        if (seenIds.has(validated.id)) {
            return this.quarantine(dir, origin, `duplicate plugin id "${validated.id}"; the copy loaded first wins`, validated.id);
        }
        seenIds.add(validated.id);

        return { id: validated.id, manifest: validated, dir, origin, status: 'discovered' };
    }

    /**
     * The plugin entry path declared by a directory's package.json, or
     * `undefined` when this is not a plugin directory. Throws only when the
     * package.json exists but cannot be read or parsed.
     */
    private async readEntryPath(dir: string): Promise<string | undefined> {
        let raw: string;
        try {
            raw = await readFile(resolve(dir, 'package.json'), 'utf8');
        } catch {
            // No package.json at all: an ordinary directory, not a failed plugin.
            return undefined;
        }

        const parsed = JSON.parse(raw) as PluginPackageJson;
        const entry = parsed.deadair?.plugin;
        if (entry === undefined) return undefined;
        if (typeof entry !== 'string' || entry.length === 0) {
            throw new Error('"deadair.plugin" must be a non-empty relative path to the plugin entry');
        }
        return entry;
    }

    /** A `failed` record: error text, no manifest, no instance. */
    private quarantine(dir: string, origin: PluginOrigin, error: string, id?: string): PluginRecord {
        return { id: id ?? basename(dir), dir, origin, status: 'failed', error };
    }
}

File: apps/desktop/src/MaroonedSoftware.Deadair.Desktop.Core/Plugins/PluginLoader.cs
using System.Globalization;
using MaroonedSoftware.Deadair.Desktop.PluginSdk;
using MaroonedSoftware.Deadair.Desktop.PluginSdk.Output;

namespace MaroonedSoftware.Deadair.Desktop.Core.Plugins;

/// <summary>
/// Finds plugins and builds them, and answers with a record whatever happens.
/// </summary>
/// <remarks>
/// <para>
/// <b>Nothing here throws.</b> Every input is a directory somebody else wrote, so a missing file, a
/// stray comma, a renamed class and an exception from a constructor are all ordinary and all end the
/// same way: a row on the settings page with a sentence on it. A loader that threw would put the
/// app's startup at the mercy of a plugin nobody has even enabled.
/// </para>
/// <para>
/// The two steps are separate on purpose. <see cref="Discover"/> reads manifests and nothing else,
/// so a disabled plugin can be listed and configured with none of its code loaded, and the
/// compatibility check happens before an assembly is opened rather than after.
/// </para>
/// </remarks>
public sealed class PluginLoader(PluginDirectories directories)
{
    private readonly Dictionary<string, PluginLoadContext> _contexts = new(StringComparer.Ordinal);

    /// <summary>
    /// Every plugin candidate in both directories, bundled first.
    /// </summary>
    /// <remarks>
    /// Bundled first because a duplicate id is decided by who was seen first, and a plugin shipped
    /// with the app should not be replaceable by dropping a folder of the same name into a directory
    /// the app does not control.
    /// </remarks>
    public IReadOnlyList<PluginRecord> Discover()
    {
        var records = new List<PluginRecord>();
        var seen = new HashSet<string>(StringComparer.Ordinal);

        foreach (var (directory, origin) in new[]
        {
            (directories.Bundled, PluginOrigin.Bundled),
            (directories.User, PluginOrigin.User),
        })
        {
            foreach (var candidate in Candidates(directory))
            {
                var record = Read(candidate, origin);

                if (!seen.Add(record.Id))
                {
                    record = record with
                    {
                        Status = PluginStatus.Failed,
                        Error = string.Create(CultureInfo.InvariantCulture, $"""another plugin is already installed as "{record.Id}", and the copy found first is the one being used."""),
                    };
                }

                records.Add(record);
            }
        }

        return records;
    }

    /// <summary>
    /// Loads the assembly and builds the entry type. Answers a record carrying the instance, or one
    /// carrying why not.
    /// </summary>
    /// <remarks>
    /// Building is separate from starting: <see cref="IDeadairPlugin.InitializeAsync"/> is the
    /// caller's to run, under its own deadline, because a plugin that hangs on start is a different
    /// problem from one that cannot be constructed and wants a different sentence.
    /// </remarks>
    public PluginRecord Activate(PluginRecord discovered)
    {
        ArgumentNullException.ThrowIfNull(discovered);

        if (discovered.Manifest is not { } manifest)
        {
            return discovered;
        }

        var assemblyPath = Path.Combine(discovered.Directory, manifest.Entry.Assembly);

        if (!File.Exists(assemblyPath))
        {
            return Failed(discovered, $"""{manifest.Entry.Assembly} is not in the plugin's folder; it may not have been built.""");
        }

        try
        {
            if (!_contexts.TryGetValue(discovered.Directory, out var context))
            {
                context = new PluginLoadContext(assemblyPath);
                _contexts[discovered.Directory] = context;
            }

            var assembly = context.LoadFromAssemblyPath(assemblyPath);
            var type = assembly.GetType(manifest.Entry.Type, throwOnError: false);

            if (type is null)
            {
                return Failed(discovered, $"""{manifest.Entry.Assembly} has no type called "{manifest.Entry.Type}"; a rename would do this.""");
            }

            if (!typeof(IDeadairPlugin).IsAssignableFrom(type))
            {
                return Failed(discovered, $"""{manifest.Entry.Type} does not implement {nameof(IDeadairPlugin)}.""");
            }

            if (Activator.CreateInstance(type) is not IDeadairPlugin instance)
            {
                return Failed(discovered, $"""{manifest.Entry.Type} could not be built; it needs a public constructor taking nothing.""");
            }

            // Checked here rather than trusted at the point of use, so a mismatch is one sentence at
            // load rather than a plugin that is listed as working and then answers nothing when the
            // output picker asks it for devices.
            if (discovered.Declares(PluginCapabilities.OutputTarget) && instance is not IOutputTargetProvider)
            {
                return Failed(discovered, $"""{manifest.Entry.Type} declares "{PluginCapabilities.OutputTarget}" but does not implement {nameof(IOutputTargetProvider)}.""");
            }

            return discovered with { Status = PluginStatus.Active, Instance = instance };
        }
        catch (Exception error) when (error is not OutOfMemoryException and not StackOverflowException)
        {
            // A plugin built against a contract this app does not have arrives here as a
            // FileLoadException or a TypeLoadException, which is what catches one that declared a
            // range it does not really need.
            return Failed(discovered, $"{manifest.Entry.Assembly} could not be loaded: {error.Message}");
        }
    }

    private static PluginRecord Failed(PluginRecord record, string error) =>
        record with { Status = PluginStatus.Failed, Error = error, Instance = null };

    /// <summary>
    /// Directories holding a manifest, sorted so that a listing is the same on two machines.
    /// </summary>
    /// <remarks>
    /// A directory without one is skipped in silence rather than reported: a plugins folder may hold
    /// notes, an archive somebody downloaded, or a <c>.DS_Store</c>, and none of those is a broken
    /// plugin.
    /// </remarks>
    private static IEnumerable<string> Candidates(string directory)
    {
        string[] subdirectories;

        try
        {
            if (!Directory.Exists(directory))
            {
                return [];
            }

            subdirectories = Directory.GetDirectories(directory);
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException)
        {
            return [];
        }

        Array.Sort(subdirectories, StringComparer.Ordinal);

        return subdirectories.Where(path => File.Exists(Path.Combine(path, PluginApi.ManifestFileName)));
    }

    private static PluginRecord Read(string directory, PluginOrigin origin)
    {
        var name = Path.GetFileName(directory);
        var reading = PluginManifestReader.ReadFile(Path.Combine(directory, PluginApi.ManifestFileName));

        if (reading.Manifest is not { } manifest)
        {
            return new PluginRecord
            {
                Id = name,
                Directory = directory,
                Origin = origin,
                Status = PluginStatus.Failed,
                Error = reading.Problem,
            };
        }

        var problem = ApiVersionRange.TryParse(manifest.ApiVersion, out var range);

        if (problem is not null)
        {
            return new PluginRecord
            {
                Id = manifest.Id,
                Directory = directory,
                Origin = origin,
                Status = PluginStatus.Failed,
                Manifest = manifest,
                Error = problem,
            };
        }

        if (!range!.Allows(PluginApi.Version))
        {
            return new PluginRecord
            {
                Id = manifest.Id,
                Directory = directory,
                Origin = origin,
                Status = PluginStatus.Failed,
                Manifest = manifest,
                Error = string.Create(CultureInfo.InvariantCulture, $"""it needs plugin API "{range.Text}" and this app has {PluginApi.Version}."""),
            };
        }

        return new PluginRecord
        {
            Id = manifest.Id,
            Directory = directory,
            Origin = origin,
            Status = PluginStatus.Discovered,
            Manifest = manifest,
        };
    }
}
